tests/acceptance/01_vars/02_functions/string.cf

```diff
@@ -0,0 +1,49 @@
+body common control
+{
+  bundlesequence => { "init", "check" };
+}
+
+bundle agent init
+{
+  classes:
+    "ExistingClass";
+
+  vars:
+    "a_int" int => "1k";
+    "a_real" real => "6.066";
+    "a_string" string => "Really interesting string";
+    "a_data_container" data => '{"foo": "bar"}';
+    "a_data_container_2" data => '{"list": [1, 2, 3, 4], "obj": {"key": ["b", "a", "r"]}}';
+
+    "string_of_int" string => string("$(a_int)");
+    "string_of_real" string => string("$(a_real)");
+    "string_of_string" string => string("$(a_string)");
+    "string_of_bool_true" string => string(and("ExistingClass"));
+    "string_of_bool_false" string => string(and("NonExistingClass"));
+    "string_of_data_container" string => string(a_data_container);
```

Comment:
How does `string("a_data_container")` behaves?

Reply:
```
vars:
  "a_data_container_3" data => '[1, 2, 3]';
  "string_of_data_container_3" string => string("a_data_container_3");
reports:
  "string of container 3 = $(string_of_data_container_3)";

### output
R: [1,2,3]
```
Is this desired behaviour?

Reply:
Not what I would expect. @nickanderson @tzz @olehermanse thoughts?

Reply:
> Not what I would expect. @nickanderson @tzz @olehermanse thoughts?

@vpodzime Why? Because there are no spaces after the commas?

Reply:
>  Because there are no spaces after the commas?

I believe thats because i used the JsonWriteCompact

Reply:
> @vpodzime Why? Because there are no spaces after the commas?

So you give the function a string. And because there's a variable with a name matching the string it gives you the string version of the variable? I don't like this. It may be OK for functions that clearly act on `slist/data` **variables**, but `string("some_string")` should just return `"some_string"`, I believe.

Reply:
I see. Sorry if I misled @KotzaBoss with my earlier suggestion. I hope I can write a better explanation here.

Based on https://tracker.mender.io/browse/CFE-3476 I think @vpodzime and others are asking for converting **anything** to a string without awareness if it's a variable or not. That's very similar to `format("%s", ...)` except for how booleans are treated, so I would mention the equivalence in the docs for `string()`. Of course, you can choose to also make `format("%s", and(...))` work the same way.

In the docs for `string()` I would also mention `format("%S", ...)` to get the "other" behavior, serializing variables by name.

Reply:
So we are interested in the InlineToJson part and not the VarName i assume?. For example a reference to json will serialize it but any "some_string" is not interpreted as anything other than a string

Reply:
> For example a reference to json will serialize it but any "some_string" is not interpreted as anything other than a string

Yes, `json_var` and `@(json_var)` should use the variable, but `"json_var"` should just be a string.

Reply:
> Yes, `json_var` and `@(json_var)` should use the variable, but `"json_var"` should just be a string.

@vpodzime AFAIK quotes are optional for strings in policy language. `json_var` and `"json_var"` are the same. `json_var` is not a reference, it is a string (that happens to be the name of a variable).

Maybe @nickanderson can clarify.

Reply:
> So you give the function a string. And because there's a variable with a name matching the string it gives you the string version of the variable? I don't like this. It may be OK for functions that clearly act on `slist/data` **variables**, but `string("some_string")` should just return `"some_string"`, I believe.

@vpodzime @KotzaBoss So, based on this and my last comment, I think `string("json_var")` and `string(json_var)` should both return the string `"json_var"`.

My understanding is that the only things that would make sense to look up the variable and serialize the JSON would be: `string(@(json_var))` (or we could create another function which takes variable names, but let's not).

Reply:
Definately, I would expect `string( @(json_var) )` to return the seralized refernece.

I don't know that I could point to anything and be able to consistently say if you should expect a bare reference to be the literal string or the expansion of a variable with the same name.

Look at all the ways `mergedata()` can work.

    bundle agent __main__
    {
      vars:
          "one" data => '{ "one": "red" }';
          "two" data => '{ "two": "blue" }';
          "merged_bare" data => mergedata( one, two);
          "merged_quote" data => mergedata( "one", "two");
          "merged_at" data => mergedata( @(one), @(two));
          "merged_quote_at" data => mergedata( "@(one)", "@(two)");
          "merged_bare_ref" data => mergedata( "@(one)", "@(two)", '{ "three": one }'); # Here, the unquoted /one/ will be the expansion of the /one/ data container, putting it under the three key.
    
      reports:
        "merged_bare: $(with)" with => string_mustache( "{{%-top-}}", merged_bare );
        "merged_quote: $(with)" with => string_mustache( "{{%-top-}}", "merged_quote" );
        "merged_at: $(with)" with => string_mustache( "{{%-top-}}", @(merged_at) );
        "merged_quote_at: $(with)" with => string_mustache( "{{%-top-}}", "@(merged_quote_at)" );
        "merged_bare_ref: $(with)" with => string_mustache( "{{%-top-}}", merged_bare_ref );
    }

    R: merged_bare: {
      "one": "red",
      "two": "blue"
    }
    R: merged_quote: {
      "one": "red",
      "two": "blue"
    }
    R: merged_at: {
      "one": "red",
      "two": "blue"
    }
    R: merged_quote_at: {
      "one": "red",
      "two": "blue"
    }
    R: merged_bare_ref: {
      "one": "red",
      "three": {
        "one": "red"
      },
      "two": "blue"
    }

Reply:
@nickanderson `mergedata()` takes variable names (or data in {} or a `@` reference). You can see from your output that quoting does not matter, just like I said. (Yes there is a "special feature" with the unquoted `one` inside the data container, but that is a separate thing, IMO).